sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRuleSets.java

```diff
@@ -141,6 +142,7 @@
 
   private static final List<RelOptRule> BEAM_CONVERTERS =
       ImmutableList.of(
+          UnionMergeRule.INSTANCE, //Added for three way union
```

Comment:
Ah. The only difference of us might just be the place to add this rule:

I added to LOGICAL_OPTIMIZATIONS above.

Reply:
If it does not work, you could 
1. re-clone a new repo of Beam
2. add UnionMergeRule
3. change `BeamCostModel.FACTORY` to `null` at https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/CalciteQueryPlanner.java#L116